File: java/08_023.java
import java.util.Arrays;
import java.util.Comparator;

/**
 * Optimizes team member allocation to cover time intervals
 */
public class IntervalCoverageOptimization {
   /**
    * Find minimum number of team members needed
    * 
    * @param intervals Array of time intervals to cover
    * @param K Length of each team member's shift
    * @return Minimum number of team members required
    */
   public static int minTeamMembers(int[][] intervals, int K) {
      // Sort intervals by start time
      Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));
      
      // Track which intervals are covered
      boolean[] covered = new boolean[intervals.length];
      int teamCount = 0;
      
      while (!allCovered(covered)) {
         teamCount++;
         
         // Find earliest uncovered interval
         int earliestUncovered = -1;
         for (int i = 0; i < intervals.length; i++) {
            if (!covered[i]) {
               earliestUncovered = i;
               break;
            }
         }
         
         if (earliestUncovered == -1) {
            break;
         }
         
         // Determine shift boundaries
         int shiftStart = intervals[earliestUncovered][0];
         int shiftEnd = shiftStart + K;
         
         // Mark all intervals covered by this shift
         for (int i = 0; i < intervals.length; i++) {
            if (!covered[i] && 
                intervals[i][0] >= shiftStart && 
                intervals[i][1] <= shiftEnd) {
               covered[i] = true;
            }
         }
      }
      
      return teamCount;
   }
   
   /**
    * Check if all intervals are covered
    */
   private static boolean allCovered(boolean[] covered) {
      for (boolean c : covered) {
         if (!c) {
            return false;
         }
      }
      return true;
   }
   
   /**
    * Test the optimization algorithm
    */
   public static void main(String[] args) {
      int[][] intervals = {
         {1, 3}, {2, 5}, {6, 8}, {8, 10}, {11, 12}
      };
      int K = 5;
      System.out.println("Minimum team members needed: " + 
            minTeamMembers(intervals, K)); // Output: 2
   }
}

File: cpp/08_023.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int minTeamMembers(std::vector<std::vector<int>>& intervals, 
                  int K) {
    // Sort intervals by start time
    std::sort(intervals.begin(), intervals.end(), 
        [](const std::vector<int>& a, 
           const std::vector<int>& b) {
            return a[0] < b[0];
        });
    
    // Keep track of covered intervals
    std::vector<bool> covered(intervals.size(), false);
    int teamCount = 0;
    
    while (std::find(covered.begin(), covered.end(), 
                    false) != covered.end()) {
        teamCount++;
        
        // Find the earliest uncovered interval
        int earliestUncovered = -1;
        for (int i = 0; i < intervals.size(); i++) {
            if (!covered[i]) {
                earliestUncovered = i;
                break;
            }
        }
        
        if (earliestUncovered == -1) {
            break;
        }
        
        // Start shift at beginning of earliest interval
        int shiftStart = intervals[earliestUncovered][0];
        int shiftEnd = shiftStart + K;
        
        // Cover all intervals that fit within this shift
        for (int i = 0; i < intervals.size(); i++) {
            if (!covered[i] && 
                intervals[i][0] >= shiftStart && 
                intervals[i][1] <= shiftEnd) {
                covered[i] = true;
            }
        }
    }
    
    return teamCount;
}

int main() {
    std::vector<std::vector<int>> intervals = 
        {{1, 3}, {2, 5}, {6, 8}, {8, 10}, {11, 12}};
    int K = 5;
    std::cout << minTeamMembers(intervals, K) << std::endl;
    return 0;
}